### ISSUE TITLE
Using test data, second image of the array "is not valid"

### ISSUE THREAD

**Issue by bokepa** [mlange-42/chrono-photo#27]
Trying to run it on Mac:

On OSx, from debug dir, and after generating test data, it fails to get second image of the folder:
./chrono-photo --debug --pattern test_data/generated/image-000*.jpg  --output out.jpg 

error: Found argument 'test_data/generated/image-00001.jpg' which wasn't expected, or isn't valid in this context

There's no problem with paths, as when I pass a single image as pattern argument, it builds OK the out.jpg file (generating same image as input, of course...):
./chrono-photo --debug --pattern test_data/generated/image-00000.jpg  --output out.jpg 

Maybe some problem with patterns and OSX?
Compilation worked without problems.

**Comment by mlange-42** [comment]
Hi @bokepa 
many thanks for testing on Mac, and for reporting this bug! Good to read that it at least compiles without problems, as I was not yet able to test that.

Reading the error message "Found argument ..." lets me suspect that the pattern is evaluated (to a list of matching paths) instead of passing it as argument. 

In the Rust docs for command line args, I read (related to Unix, but might may appy for Mac too):
> On Unix systems shell usually expands unquoted arguments with glob patterns (such as * and ?). On Windows this is not done, and such arguments are passed as-is."

Could you please **try quoting the pattern**, maybe that solves the problem? i.e. `--pattern "test_data/generated/image-000*.jpg"`

In case it does not help, did you run it from an IDE, or any other way with debug information, so you could provide the stack trace?


Many thanks!

**Comment by bokepa** [comment]
Thanks @mlange-42 , It works like a charm. 

First days on Rust, so dont'n know which ide for debugging. I'm looking for...

Now testing it with 4 frames set, but having problems with "too many open file" OS error,... limit on my old macos is 12.000, I think. But not very sure about which limit is overflowing.
 
By now, setting --slice count/100 avoid the problem.

Thank you

**Comment by mlange-42** [comment]
Thanks @bokepa , and good to know that quoting helps. I already updated the documentation.

Coming from Java/Scala, I use **IntelliJ IDEA** also for Rust, and it works fairly good. https://www.jetbrains.com/idea/download/#section=mac
("Community version" at the right)

For the "open files" problem: The number of open files (writing temp files) does not depend on the number of images, but on the image size (actually image height under default settings). As you already noticed, it can be controlled via `--slice`. The default (`rows/4`) produces `image rows/4` temp files. Which should be well below 12'000.

If you do not process thousands of images, a setting like `--slice count/100` should work well. It needs to be changed only if you observe too high RAM usage during the 2nd stage of processing (2nd progress bar).

I will test if it hurts performance to open and close the file streams repeatedly.

Regarding the **seagull tests**, I will explain the reason to XB, and ask him to forward it to you, and maybe include you in our communication if that is ok for you.

Further, do you think you could help by building new releases for publication on GitHub, as I already do with Win executables? I am also looking for someone under my connections for that, but Mac is very rare among them.

Thank you!

**Comment by mlange-42** [comment]
@bokepa with version 0.4.2 (latest on `master`), the "open files limit" problem should be solved.

Could you try with `--slice rows/1`?

Thanks!

**Comment by bokepa** [comment]
I've tried without slice parameter and with --slice rows/1, and max opened files error doesn't happens now.

**Comment by mlange-42** [comment]
@bokepa Great, thanks again!

**Comment by mlange-42** [comment]
@bokepa Just wanted to make sure that you used `cargo build --release` to build the Mac binaries for XB. Did you?

Building without `--release` enables debug information and disables compiler optimizations, which makes things by orders of magnitude slower.

Many thanks for confirmation!

**Comment by bokepa** [comment]
You are right. At first I compiled it without release option.

Day after, while trying to learn a little bit Rust and reading its
documentation I realized about the --release option. I guess XB executed
that one version. (On may old 2009 Mac everything is slow, and it was a big
difference between the debuh version and release version)

Not compiled again since 0.4.2. I've had some crash sometimes, don't know
why. I'll send you the exception throw. And not have much time to pull
again...

Thank you!


El dg., 19 d’abr. 2020, 18:09, mlange-42 <notifications@github.com> va
escriure:
